Fix CORS preflight requests

masad-frost opened the pull request:
Currently the custom CORS middleware that the server has doesn't handle `OPTIONS` requests which are used to check the CORS policy of the server.

Got rid of the custom cors middleware and express's official `cors` middleware which handles all the nitty gritty of CORS requests. I made sure I pass all current CORS header values to the `cors` middleware

We use this API at repl.it for people to search for packages and add them to their HTML. All our requests are preflighted making the API unusable for us. We circumvented the issue by removing the content type (json) header, and will probably proxy the API through our servers for extra safety until this is deployed.

Thank you!

MattIPv4 commented on app.js:
```diff
@@ -48,17 +49,20 @@ module.exports = () => {
+        origin: '*',
+        credentials: true,
+        methods: ['GET', 'PUT', 'POST', 'DELETE', 'OPTIONS'],
+        allowedHeaders: ['X-CSRF-Token', ' X-Requested-With', ' Accept', ' Accept-Version', ' Content-Length', ' Content-MD5', ' Content-Type', ' Date', ' X-Api-Version'],
```
There seems to be some extra spacing inside the strings here, is that intended?

masad-frost replied:
Oh my bad! I'll add tests for this.

MattIPv4 commented:
I wonder if it'd also be worth adding options tests to each route, just to ensure this isn't accidentally broken down the road?

masad-frost commented:
Yeah I can add exhaustive tests!

MattIPv4 commented:
Hey @masad-frost, just wanted to check in on this. If you don't have time to add these extras, I'm happy to either push to your fork, or to merge this and then make the changes after. Let me know what works best for you! :)

masad-frost commented:
Hey, sorry about this. I can do it this weekend. Feel free to merge and I'll follow up

MattIPv4 commented on tests/suite/index.js:
```diff
@@ -16,6 +16,13 @@ describe('/', () => {
         expect(response).to.have.header('Cache-Control', 'public, max-age=30672000, immutable'); // 355 days
         done();
     });
+    it('returns the correct CORS and Cache headers for OPTIONS', done => {
+        request().options('/').redirects(0).end((err, res) => {
+            expect(res).to.have.header('Access-Control-Allow-Origin', '*');
+            expect(res).to.have.header('Cache-Control', 'public, max-age=30672000, immutable'); // 355 days
+            done();
+        });
+    });
```
```suggestion
    it('returns the correct CORS headers for OPTIONS', done => {
        request().options('/').redirects(0).end((err, res) => {
            expect(res).to.have.header('Access-Control-Allow-Origin', '*');
            done();
        });
    });
```

This currently fails when tests are run, I'm guessing custom headers aren't mirrored from the GET routes, which is fine.

masad-frost replied:
Oh yeah that makes sense. We can add it, but we `Cache-Control` doesn't really make sense for `options`

drewfreyling replied:
agreed

masad-frost commented:
Updated, added to tests to all routes. Disclaimer, I used macros to add those since all the tests have the same format! So hopefully my laziness didn't make me miss anything.

MattIPv4 commented on tests/cors.js:
```diff
@@ -0,0 +1,36 @@
+const request = require('./base');
+const { corsOptions } = require('../utils/cors');
+
+const expectCORSHeaders = (response) => {
```
```suggestion
const expectCORSHeaders = response => {
```
Sorry to be pedantic, but I like to have consistency :)

masad-frost replied:
Seems like something `--fix` should fix.

MattIPv4 commented on tests/cors.js:
```diff
@@ -0,0 +1,36 @@
+    expect(response).to.have.header('Access-Control-Allow-Origin', corsOptions.origin);
+    expect(response).to.have.header('Access-Control-Allow-Credentials', corsOptions.credentials.toString());
+
+    // testing allowed methods and headers only makes sense in case of options
```
```suggestion
    // Testing allowed methods and headers only makes sense in case of options
```

MattIPv4 commented on tests/cors.js:
```diff
@@ -0,0 +1,36 @@
+        done();
+    });
+
+    it('returns the correct CORS headers for OPTIONS request', (done) => {
```
```suggestion
    it('returns the correct CORS headers for OPTIONS request', done => {
```

MattIPv4 commented on tests/suite/libraries.js:
```diff
@@ -193,16 +198,16 @@ describe('/libraries', function () {
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/libraries.js:
```diff
@@ -240,16 +245,16 @@ describe('/libraries', function () {
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/libraries.js:
```diff
@@ -276,16 +281,16 @@ describe('/libraries', function () {
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/libraries.js:
```diff
@@ -327,16 +332,16 @@ describe('/libraries', function () {
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/library.js:
```diff
@@ -136,16 +141,16 @@ describe('/libraries/:library/:version', () => {
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/library.js:
```diff
@@ -164,16 +169,16 @@ describe('/libraries/:library/:version', () => {
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/library.js:
```diff
@@ -247,16 +252,16 @@ describe('/libraries/:library', () => {
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/library.js:
```diff
@@ -278,16 +283,16 @@ describe('/libraries/:library', () => {
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/library.js:
```diff
@@ -322,16 +327,16 @@ describe('/libraries/:library', () => {
             });
-            it('returns the correct CORS and Cache headers', done => {
-                expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+            it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/tutorials.js:
```diff
@@ -148,16 +153,16 @@ describe('/libraries/:library/tutorials/:tutorial', () => {
                 });
-                it('returns the correct CORS and Cache headers', done => {
-                    expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+                it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/tutorials.js:
```diff
@@ -180,16 +185,16 @@ describe('/libraries/:library/tutorials/:tutorial', () => {
                 });
-                it('returns the correct CORS and Cache headers', done => {
-                    expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+                it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/tutorials.js:
```diff
@@ -211,16 +216,16 @@ describe('/libraries/:library/tutorials/:tutorial', () => {
                 });
-                it('returns the correct CORS and Cache headers', done => {
-                    expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+                it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/tutorials.js:
```diff
@@ -246,16 +251,16 @@ describe('/libraries/:library/tutorials/:tutorial', () => {
                 });
-                it('returns the correct CORS and Cache headers', done => {
-                    expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+                it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/tutorials.js:
```diff
@@ -274,16 +279,16 @@ describe('/libraries/:library/tutorials/:tutorial', () => {
                 });
-                it('returns the correct CORS and Cache headers', done => {
-                    expect(response).to.have.header('Access-Control-Allow-Origin', '*');
+                it('returns the correct Cache headers', done => {
```
CORS test missing

MattIPv4 commented on tests/suite/update.js:
```diff
@@ -1,9 +1,11 @@
 // This endpoint is a prod-only endpoint for monitoring the API auto-update job by maintainers
 describe('/update', () => {
+    const path = '/update';
     const test = () => request().get('/update');
```
```suggestion
    const test = () => request().get(path);
```

MattIPv4 commented:
Hey @masad-frost,

Would you like me to handle rebasing this and adding in the missing CORS tests, so that we can get this merged?

masad-frost commented:
On my list today! Thanks for the reminder

On Wed, Jun 3, 2020, 3:35 AM Matt (IPv4) Cowley <notifications@github.com>
wrote:

> Hey @masad-frost <https://github.com/masad-frost>,
>
> Would you like me to handle rebasing this and adding in the missing CORS
> tests, so that we can get this merged?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cdnjs/api-server/pull/17#issuecomment-638110447>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPG2LGD6FGCXIEIAXVBI5LRUYRPZANCNFSM4NET66SA>
> .
>

MattIPv4 commented:
Hey @masad-frost -- apologies for pinging again but I just wanted to check in on this as it sounds this was originally created due to this causing issues for you in production, so we'd love to get this shipped to help y'all out.

If you don't think you'll have the time to get this ready to be landed I'm happy to take over and add the missing tests + rebasing :)

masad-frost commented:
Oh boy, my bad. On it!

masad-frost commented:
Should be good to go now!

masad-frost commented:
Thanks for pulling this in and maintaining cdnjs!

No rush, we're proxying requests through our domain right now. Here it is in action:
![cdnjscool](https://user-images.githubusercontent.com/1994028/84960284-15881e00-b0b6-11ea-987e-ac76de82949e.gif)

MattIPv4 commented:
Heh, awesome! Love seeing easy to use UI for bring cdnjs libraries in.

My one suggestion would be to encourage you to default to including the integrity attribute on the asset includes (CSS & JS) — SRI data should be available in the library API responses :)

masad-frost commented:
Sweet, will add it.
